flink-tests/src/test/java/org/apache/flink/test/checkpointing/ZooKeeperHighAvailabilityITCase.java

```diff
@@ -0,0 +1,387 @@
+			},
+			deadline,
+			TestingUtils.defaultExecutor());
+		assertEquals(JobStatus.RESTARTING, jobStatusFuture.get());
+
+		jobStatusFuture = FutureUtils.retrySuccessful(
+			new Callable<Future<JobStatus>>() {
+				@Override
+				public Future<JobStatus> call() {
+					return getJobStatus(jobManager, jobID, TEST_TIMEOUT);
+				}
+			},
+			new FilterFunction<JobStatus>() {
+				@Override
+				public boolean filter(JobStatus jobStatus) {
+					return jobStatus == JobStatus.FAILING;
+				}
+			},
+			deadline,
+			TestingUtils.defaultExecutor());
+		assertEquals(JobStatus.FAILING, jobStatusFuture.get());
+
+		// move back the HA directory so that the job can restore
+		CheckpointBlockingFunction.afterMessWithZooKeeper.set(true);
+
+		files = movedCheckpointLocation.listFiles();
+		assertNotNull(files);
+		for (File file : files) {
+			if (file.getName().startsWith("completedCheckpoint")) {
+				assertTrue(file.renameTo(new File(haStorageDir, file.getName())));
+			}
+		}
+
+		// now the job should be able to go to RUNNING again and then eventually to FINISHED
+		jobStatusFuture = FutureUtils.retrySuccessful(
+			new Callable<Future<JobStatus>>() {
+				@Override
+				public Future<JobStatus> call() {
+					return getJobStatus(jobManager, jobID, TEST_TIMEOUT);
+				}
+			},
+			new FilterFunction<JobStatus>() {
+				@Override
+				public boolean filter(JobStatus jobStatus) {
+					return jobStatus == JobStatus.FINISHED;
+				}
+			},
+			deadline,
+			TestingUtils.defaultExecutor());
+		assertEquals(JobStatus.FINISHED, jobStatusFuture.get());
+
+		// make sure we saw a successful restore
+		successfulRestoreLatch.await();
+
+		assertThat("We saw illegal restores.", CheckpointBlockingFunction.illegalRestores.get(), is(0));
+	}
+
+	/**
+	 * Requests the {@link JobStatus} of the job with the given {@link JobID}.
+	 */
+	private Future<JobStatus> getJobStatus(
+		final ActorGateway jobManager,
+		final JobID jobId,
+		final FiniteDuration timeout) {
+
+		scala.concurrent.Future<Object> response =
+			jobManager.ask(JobManagerMessages.getRequestJobStatus(jobId), timeout);
+
+		FlinkFuture<Object> flinkFuture = new FlinkFuture<>(response);
+
+		return flinkFuture.thenApply(new ApplyFunction<Object, JobStatus>() {
+			@Override
+			public JobStatus apply(Object value) {
+				if (value instanceof JobManagerMessages.CurrentJobStatus) {
+					return ((JobManagerMessages.CurrentJobStatus) value).status();
+				} else if (value instanceof JobManagerMessages.JobNotFound) {
+					throw new RuntimeException(
+						new IllegalStateException("Could not find job with JobId " + jobId));
+				} else {
+					throw new RuntimeException(
+						new IllegalStateException("Unknown JobManager response of type " + value.getClass()));
+				}
+			}
+		});
+	}
+
+	private static class UnboundedSource implements SourceFunction<String> {
+		private boolean running = true;
+
+		@Override
+		public void run(SourceContext<String> ctx) throws Exception {
+			while (running) {
```

Comment:
`while (running && !CheckpointBlockingFunction.afterMessWithZooKeeper.get())`?

Reply:
I had this in an earlier version, but then I did some messing around. Fixing